### PR TITLE
Breakout DShot driver header file, use static_casts, add const specifiers, format and standardize throughout the driver.

### PR DESCRIPTION
**Pull request by mcsauder**
Hello,

This PR aims to improve readability in the dshot driver source file through the following modifications:
 - A header file has been added and dshot driver specific #defines moved out of `drv_pwm_output.h` to `dshot.h`,
 - C-style casts have been replaced with static_casts,
 - Method naming convention (non-inherited methods) and single line comment style and formatting have been standardized for self consistency within the class,
 - Additional `const` specifiers have been added to non-inherited method definitions/declarations where applicable.

**Test data / coverage**
This PR has been tested with a Holybro pixhawk 4 on an X-class racing quad with four APD 80F3 dshot ESCs.

Let me know if you have any questions on this PR. Thanks!

-Mark

### CODE REVIEW

**Comment by dagar** [comment]
Minor formatting complains from CI. https://github.com/PX4/Firmware/pull/15908/checks?check_run_id=1222963359

Screen Shot 2020-10-07 at 7 47 02 PM

**Comment by dagar** [comment]
This might be controversial bikeshedding, but if you're going to rename why not name it the same as what's actually inside (ie DShotOutput.hpp). File naming is one area where I disagree with the google c++ style guide.

**Comment by mcsauder** [comment]
> This might be controversial bikeshedding, but if you're going to rename why not name it the same as what's actually inside (ie DShotOutput.hpp). File naming is one area where I disagree with the google c++ style guide.

I wholeheartedly agree with you! Thanks for bringing it to the surface, I'll tidy that up!

**Comment by mcsauder** [comment]
(As a footnote, the motivation behind this PR is that APD's latest ESC firmware is incompatible with the latest stable release, v1.11.  I am working on tracking down why v1.11 doesn't work but the current master branch does.)

**Comment by mcsauder** [comment]
On the placement of the `*`, it's funny to me but not worth wasting too much time.  My thought is, we are programming in a strongly typed language on purpose, (otherwise we would be Ardupilot ;) JK), so, what exactly is the actual type of `Telemetry *telemetry`,... The type is a *pointer* of type `Telemetry`, so isn't that a `Telemetry*` type? :)  Done with that thought. :)

File: src/drivers/dshot/DShot.cpp [hunk 1]
```diff
@@ -668,25 +470,31 @@ void DShotOutput::retrieveAndPrintESCInfoThreadSafe(int motor_index)
 	DShotTelemetry::decodeAndPrintEscInfoPacket(output_buffer);
 }
 
-int DShotOutput::requestESCInfo()
+int
+DShot::request_esc_info()
```
Comment (dagar) [inline src/drivers/dshot/DShot.cpp:474]: I've been doing return type on the same line as the function name (https://google.github.io/styleguide/cppguide.html#Function_Declarations_and_Definitions). 

Do you have a strong preference?

Comment (mcsauder) [reply src/drivers/dshot/DShot.cpp:474]: I can make them all on the same line, in current master they are a mix of both styles.

Comment (mcsauder) [reply src/drivers/dshot/DShot.cpp]: Done.

**Comment by mcsauder** [comment]
SITL test failures are unrelated to this PR: 
image

**Comment by mcsauder** [comment]
Flight tested with dshot 1200: https://review.px4.io/plot_app?log=b5a5e6fb-38d4-4ef9-9d10-2eab4f75c6ee

**Comment by dagar** [comment]
CI is failing rather unhelpfully and I'm not sure why.

Screen Shot 2020-12-01 at 4 54 08 PM


Can you rebase on current master and force push?

File: src/drivers/dshot/DShot.cpp [hunk 2]
```diff
@@ -308,9 +97,9 @@ DShotOutput::set_mode(Mode mode)
 	 * listening and mixing; the mode just selects which of the channels
 	 * are presented on the output pins.
 	 */
-	switch (mode) {
+	switch (new_mode) {
```
Comment (bkueng) [inline src/drivers/dshot/DShot.cpp:100]: Please keep this in sync with https://github.com/PX4/PX4-Autopilot/blob/master/src/drivers/pwm_out/PWMOut.cpp#L103

File: src/drivers/dshot/DShot.cpp [hunk 3]
```diff
@@ -489,27 +277,26 @@ DShotOutput::task_spawn(int argc, char *argv[])
...
+			     const uint32_t edge_state, const uint32_t overflow)
 {
-	fprintf(stdout, "DShot: Capture chan:%d time:%lld state:%d overflow:%d\n", chan_index, edge_time, edge_state, overflow);
+	fprintf(stdout, "DShot: Capture chan:%d time:%lld state:%d overflow:%d\n", channel_index, edge_time, edge_state,
+		overflow);
 }
 
-void
-DShotOutput::update_dshot_out_state(bool on)
+void DShot::update_dshot_out_state(const bool state)
```
Comment (bkueng) [inline src/drivers/dshot/DShot.cpp:294]: How is `state` clearer than `on`?

File: src/drivers/dshot/DShot.h [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+	enum class DShotConfig {
+		Disabled  = 0,
+		DShot150  = 150,
+		DShot300  = 300,
+		DShot600  = 600,
+		DShot1200 = 1200,
+	};
+
+	struct Command {
+		const char *name{nullptr};
```
Comment (bkueng) [inline src/drivers/dshot/DShot.h:166]: Why did you add this?

File: src/drivers/dshot/DShot.h [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+#  error "board_config.h needs to define BOARD_HAS_PWM"
+#endif
+
+/** Dshot PWM frequency, Hz */
+static constexpr unsigned int DSHOT150  =  150000u;
+static constexpr unsigned int DSHOT300  =  300000u;
+static constexpr unsigned int DSHOT600  =  600000u;
+static constexpr unsigned int DSHOT1200 = 1200000u;
+
+static constexpr int DSHOT_DISARM_CMD   = dshot_command_t::DShot_cmd_motor_stop;
```
Comment (bkueng) [inline src/drivers/dshot/DShot.h:57]: This is confusing, it's not the same thing. Command != throttle value. The disarmed value is a throttle, and if the output requests that, it get's changed to a stop command (see `updateOutputs`).
So I'd rather keep the previous implementation, which reflected that correctly (with the +1, which might still be a separate define).
```
static constexpr int DSHOT_DISARMED_THROTTLE   = 0;
static constexpr int DSHOT_MIN_THROTTLE = DSHOT_DISARMED_THROTTLE + 1;
static constexpr int DSHOT_MAX_THROTTLE = 1999;
```

File: src/drivers/dshot/DShot.cpp (hunk 2 above)
Comment (mcsauder) [reply src/drivers/dshot/DShot.cpp:100]: OK. Thanks!

File: src/drivers/dshot/DShot.cpp (hunk 3 above)
Comment (mcsauder) [reply src/drivers/dshot/DShot.cpp:294]: what is the name of the method?

File: src/drivers/dshot/DShot.h (hunk 5 above)
Comment (mcsauder) [reply src/drivers/dshot/DShot.h:57]: Cool. I can do that!

File: src/drivers/dshot/DShot.cpp (hunk 3 above)
Comment (mcsauder) [reply src/drivers/dshot/DShot.cpp:294]: so that this could be condensed: https://github.com/PX4/PX4-Autopilot/blob/25c113a5273ab315c511773b23ef81e2a93f0026/src/drivers/dshot/dshot.cpp#L1408

File: src/drivers/dshot/DShot.h (hunk 4 above)
Comment (mcsauder) [reply src/drivers/dshot/DShot.h:166]: so that this could be condensed: https://github.com/PX4/PX4-Autopilot/blob/25c113a5273ab315c511773b23ef81e2a93f0026/src/drivers/dshot/dshot.cpp#L1408

File: src/drivers/dshot/DShot.h (hunk 5 above)
Comment (mcsauder) [reply src/drivers/dshot/DShot.h:57]: However, given the dshot protocol, 0 is a command, not a valid throttle... so I'd like to keep the name as CMD.

File: src/drivers/dshot/DShot.h (hunk 4 above)
Comment (bkueng) [reply src/drivers/dshot/DShot.h:166]: Please don't. `Command` is used internally, and might get filled in in future from other sources as well. `VerbCommand` is specifically for CLI, so the separation is on purpose.

File: src/drivers/dshot/DShot.cpp (hunk 3 above)
Comment (bkueng) [reply src/drivers/dshot/DShot.cpp:294]: That doesn't make it clearer, `state` alone could also refer to e.g. 3D mode, armed, etc (`on` was more specific). Yes this might be nitpicky, but as you want to improve it, please go all the way. `enable_outputs(bool enabled)` would be an option.

File: src/drivers/dshot/DShot.h (hunk 4 above)
Comment (mcsauder) [reply src/drivers/dshot/DShot.h]: Removed!  Thanks @bkueng !

File: src/drivers/dshot/DShot.cpp (hunk 3 above)
Comment (mcsauder) [reply src/drivers/dshot/DShot.cpp:294]: Done! Thanks!

**Comment by mcsauder** [comment]
Flight tested today with current master merged in as well as board configs and an automated test routine I wrote: https://review.px4.io/plot_app?log=898b6ac7-cf8a-41d0-abdc-9675b1a6a4f9
Everything worked great.

File: src/drivers/dshot/DShot.cpp [hunk 6]
```diff
@@ -31,257 +31,47 @@
...
-DShotOutput::DShotOutput() :
+DShot::DShot() :
 	CDev("/dev/dshot"),
-	OutputModuleInterface(MODULE_NAME, px4::wq_configurations::hp_default),
-	_cycle_perf(perf_alloc(PC_ELAPSED, MODULE_NAME": cycle"))
+	OutputModuleInterface(MODULE_NAME, px4::wq_configurations::hp_default)
 {
-	_mixing_output.setAllDisarmedValues(DISARMED_VALUE);
-	_mixing_output.setAllMinValues(DISARMED_VALUE + 1);
+	_mixing_output.setAllDisarmedValues(DSHOT_DISARM_CMD);
```
Comment (bkueng) [inline src/drivers/dshot/DShot.cpp:43]: Can you change this to `DSHOT_DISARM_VALUE` / `DSHOT_DISARM_THROTTLE` to avoid potential confusion?

Comment (mcsauder) [reply src/drivers/dshot/DShot.cpp]: Certainly. Done. Rebased against current master, commits squashed and pushed up.

File: src/drivers/drv_pwm_output.h [hunk 7]
```diff
@@ -424,20 +416,20 @@ __EXPORT extern servo_position_t up_pwm_servo_get(unsigned channel);
...
  */
 __EXPORT extern int up_dshot_init(uint32_t channel_mask, unsigned dshot_pwm_freq);
 
 /**
  * Set the current dshot throttle value for a channel (motor).
  *
  * @param channel	The channel to set.
- * @param throttle	The output dshot throttle value in [0, 1999 = DSHOT_MAX_THROTTLE].
- * @param telemetry If true, request telemetry from that motor
+ * @param throttle	The output dshot throttle value in [0 = DSHOT_DISARM_CMD, 1999 = DSHOT_MAX_THROTTLE].
```
Comment (bkueng) [inline src/drivers/drv_pwm_output.h:431]: One last thing: you forgot to update this one. However I'd rather not mention it, as 0 isn't actually set as throttle (`up_dshot_motor_command` is used instead for disarming).
```suggestion
 * @param throttle	The output dshot throttle value in [0, 1999 = DSHOT_MAX_THROTTLE].
```

Comment (mcsauder) [reply src/drivers/drv_pwm_output.h]: Thanks! Nice catch!  I added `1 = DSHOT_MIN_THROTTLE` as well.

**Review by bkueng — APPROVED** [review]

**Comment by mcsauder** [comment]
Thanks @bkueng !
